design_notes/XDN03 - xlang Type System.md

```diff
@@ -15,3 +15,142 @@ status: draft
+
+## Structs
+
+Structs are a record type with one or more fields. Structs are always passed and returned by value.
+Struct fields may only be enums, structs and fundamental types (including strings).
+
+Structs must have at least one field. All of a struct's fields are public.
+
+Structs cannot be generic or parameterized.
+
+Structs are not versionable. Once they have been defined, they may never be changed.
+
+## Object Members
+
+### Methods
+
+### Properties
+
+### Events
+
+## Interfaces
+
+> Note, an xlang's type system interface is a different concept from
+[COM Interfaces](XDN05%20-%20xlang%20Binary%20Interface.md#com-interfaces) in xlang's
+[ABI](XDN05%20-%20xlang%20Binary%20Interface.md).
+So while an xlang ABI interface must inherit from
+[IXlangObject](XDN05%20-%20xlang%20Binary%20Interface.md#ixlangobject), interfaces in the xlang
+type system do not. Inheriting from IXlangObject is an implementation detail that is not directly
+surfaced in xlang's type system.
+
+An interface is a type that acts as a contract that can be implemented by multiple classes. An
+interface specifies the signatures for a group of object members (aka methods, properties and
+events) but does not specify an implementation for those object members.
+
+Interfaces can be implemented by multiple different classes. Each class implementation of a given
+interface's object members is unrelated to any other class's implementation of the same interface.
+
+Interfaces may optionally inherit from a single other interfaces. All of the type members of the
+base interface are considered part of the derived interface as well. Any class implementing an
+interface that uses inheritance must implement all of the type members specified by both the base
+and derived interfaces.
+
+> Note, WinRT interfaces support inheritance from multiple other interfaces. This approach leads to
+inefficiencies at the ABI layer that xlang wishes to avoid by constraining interfaces to single
+inheritance. This decision may be revisited as the xlang project progresses.
+
+Interfaces are not versionable. Once they have been defined, they may never be changed.
+
+### Parameterized Interfaces
+
+xlang interfaces support type parameterization. Parameterized interfaces specify one or more type
+parameters that can be used as parameter types or return values of type members.
+
+Parameterized interfaces can inherit from both parameterized and non-parameterized interfaces.
+When inheriting from a parameterized interface, the type parameters of the base interface must
+be specified, though they may be specified in terms of the derived interface's type parameters.
+
+Non-parameterized interfaces can only inherit from concrete interfaces. This includes
+non-parameterized interfaces as  well as parameterized interfaces where all of the type parameters
+have been fully specified.
+
+Array types may not be used as type parameters.
```

Comment:
There are no array types. There are array parameters, but they are not types (in metadata).

Reply:
Nice, I like that way of phrasing it.

Reply:
cutting the array types line. When I get to writing up the section on parameters, I'll talk about arrays there.